packages/plc/src/server/index.ts

```diff
@@ -6,45 +6,76 @@ import 'express-async-errors'
 
 import express from 'express'
 import cors from 'cors'
+import http from 'http'
 import { Database } from './db'
 import * as error from './error'
-import router from './routes'
-import { Locals } from './locals'
-import { loggerMiddleware, logger } from './logger'
+import createRouter from './routes'
+import { loggerMiddleware } from './logger'
+import AppContext from './context'
+import { createHttpTerminator } from 'http-terminator'
+
 export * from './db'
+export * from './context'
 
-export type App = express.Application
+export class PlcServer {
+  public ctx: AppContext
+  public app: express.Application
+  public server?: http.Server
 
-export const server = (db: Database, port?: number, _version?: string) => {
-  const version = _version || '0.0.0'
-  const app = express()
-  app.use(express.json({ limit: '100kb' }))
-  app.use(cors())
+  constructor(opts: { ctx: AppContext; app: express.Application }) {
+    this.ctx = opts.ctx
+    this.app = opts.app
+  }
 
-  app.use(loggerMiddleware)
+  static create(opts: {
+    db: Database
+    port?: number
+    version?: string
+  }): PlcServer {
+    const app = express()
+    app.use(express.json({ limit: '100kb' }))
+    app.use(cors())
 
-  const locals: Locals = {
-    logger,
-    version,
-    db,
-  }
+    app.use(loggerMiddleware)
 
-  app.locals = locals
+    const ctx = new AppContext({
+      db: opts.db,
+      version: opts.version || '0.0.0',
+      port: opts.port,
+    })
 
-  app.use((req, res, next) => {
-    const reqLocals: Locals = {
-      ...locals,
-      logger: req.log, // This logger is request-specific
-    }
-    res.locals = reqLocals
-    next()
-  })
+    app.use('/', createRouter(ctx))
+    app.use(error.handler)
 
-  app.use('/', router)
-  app.use(error.handler)
+    return new PlcServer({
+      ctx,
+      app,
+    })
+  }
 
-  const listener = app.listen(port)
-  return { app, listener }
+  async start(): Promise<http.Server> {
+    const server = this.app.listen(this.ctx.port)
+    this.server = server
+    return new Promise((resolve, reject) => {
+      server.on('listening', () => {
+        resolve(server)
+      })
+      server.on('error', (err) => {
+        reject(err)
+      })
+    })
+  }
+
+  async destroy(force = false) {
+    if (this.server) {
+      const terminator = createHttpTerminator({
```

Comment:
Echoing https://github.com/bluesky-social/atproto/pull/424#discussion_r1050882244: it's possible we should create the terminator before starting to serve requests.  I have a question out to them about that usage detail.